feat: new options for the vertical space tool

Doublonmousse opened the pull request:
Adds the following options to the vertical space bar
![image](https://github.com/flxzt/rnote/assets/115779707/1406d525-831e-4f68-8c2d-b2d3deeb6703)

For #1048 and #878

- [x] make the two last options work
- [x] adapt the gui part of the tool to only go as far as the page widths selected

Doublonmousse committed:
new options for the vertical space tool (wip)

Doublonmousse committed:
make the two vertical/horizontal options work

Doublonmousse committed:
restrict the width of the vertical space tool to the page

Doublonmousse committed:
fix for incorrect line position

Doublonmousse commented:
I think I've done everything except doing the more complex tool with borders that can be edited. I wonder if it's not better to do this later in a separate PR.

flxzt commented:
I like this a lot from a functionality POV.

Some suggestions for improvements in the UI:
I don't like the location of the added options.
It would be better to place this right next to the tool
and rearrange the tools sidebar with visible separators between the individual tools.
This would also be consistent to the other sidebar pages.

There are also some minor inconsistencies like the option button size which I think should be the smaller variant.

Also: the option names could improved imho. What about instead of "Shape Configuration" -> " "Vertical-Space Tool Options", "Respect Vertical/Horizontal Borders" -> "Limit movement to Vertical/Horizontal Page Borders" ?

Doublonmousse committed:
ui tweaks and renaming

- renaming to be more clear
- change the ui (so that the setting button is closer to the tool it's associated with + linked list to better indicate the choice of tools)

Doublonmousse commented:
<img width="435" alt="Screenshot 2024-08-21 at 00 17 29" src="https://github.com/user-attachments/assets/93866b94-b1ab-443b-8922-c016e5f81ba9">

I've went for the linked style of other toolspages for now.. I think it's a little better at showing that there are three different tools than before.

It doesn't really fix the fact that the settings icon isn't related to the vertical space tool though.
Ideally I'd like to keep the linked aspect of the three tools whilst making the settings icon appear tied to the vertical space tool

flxzt commented:
What do you think about something like this?

![grafik](https://github.com/user-attachments/assets/72589f33-2885-4cfe-8ba7-470bb90d2a5a)

Possibly with separators between the individual tools as well

flxzt committed:
improv: some UI improvements in the tools sidebar

Doublonmousse commented:
Yes, that works well !

flxzt commented:
alright, I'll add my commit then.

Oh and I am not sure the force snapping option is actually needed, since it is already covered by the "snap positions" option. It could be confusing when deactivating one of them, why it seemingly is not respected

Doublonmousse commented:
Yeah, makes sense.
It's more of a habit I took since https://github.com/flxzt/rnote/discussions/721

I think it makes sense to have a second snap mode where displacement are multiple of the grid size as opposed to snapping elements to the grid (usecase : select some text and change its position so that it stays aligned to the grid). But, same thing as before, it makes more sense to do that as a global setting than this

flxzt commented:
Yeah. I am a bit concerned for that option because of feature creep of more "niche" use cases.

Can be a follow-up?

Doublonmousse commented:
Yes, I agree it makes sense to remove the "force snapping" option from the current PR.

And I'll see if I want to do a follow up for snapping options later (but there is probably less niche things to add to snapping before I do that)

Doublonmousse committed:
remove superfluous "force snapping" option

flxzt commented on crates/rnote-compose/src/verticaltoolconfig.rs:
```diff
@@ -0,0 +1,10 @@
+/// configuration for the vertical tool
+pub struct VerticalToolConfig {
+    /// horizontal limit
+    pub horizontal_border: bool,
```
I'd rename these fields to `limit_movement_<horizontal/vertical>_borders`

flxzt commented on crates/rnote-engine/src/store/stroke_comp.rs:
```diff
@@ -640,6 +640,64 @@ impl StrokeStore {
             .collect::<Vec<StrokeKey>>()
     }
 
+    pub(crate) fn keys_between(
```
Can this function be split up into two separate ones that are only responsible for x or y? And let the keys be supplied by the user for composing them together to get the same result.

This would get rid of the complexity (match statement) in the function body and would make it more re-usable.

Doublonmousse replied:
I've not done exactly that for now but I got rid of the match statement.
It's replaced with multiple filters using the bools for the state of the verticalspace tool to skip unneeded selection using the lazily evaluated or operator.

I've not tried to chain the filters, that'd need for the `stroke_components` to be public (or at least the iterator over them through a getter method) + two external functions defined on the `impl Iterator<Item  = (Strokekey,...)>` (can't be a method on self).

But at that point, I can also get the iterator from the new getter, and do the filters inside `toolspage.rs` (and maybe lazily evaluate some of the values I'm for now passing as parameters to `keys_between`)

flxzt replied:
Check out `filter_keys_intersecting_bounds()`, that's what I had in mind.

Wouldn't something like this be possible?

```rust
let keys = store.keys_unordered();
if limit_movement_vertical_border {
    let filtered = store.keys_between_vertical_position(&keys);
    // ...
} else {
    let filtered = store.keys_below_y()
    // ...
}

flxzt replied:
But I won't block merging this if it isn't done this way, so your call

Doublonmousse replied:
I'll try to do the method used in `filter_keys_intersecting_bounds()` and see how that compares. 

One question I have for this (and other related operations), is that I'm not sure we're using the Rtree everywhere for queries (there's a `keys_in_bounds` method that seems like a good candidate for doing this instead, unless there's issue if we have semi infinite bounds ?)

flxzt replied:
yeah that's true `keys_in_bounds` could be used instead!

flxzt replied:
Ok, this is fine for me now

flxzt commented on crates/rnote-ui/data/ui/penssidebar/._toolspage.ui:
I think this file was added by mistake

flxzt commented on crates/rnote-ui/src/penssidebar/toolspage.rs:
```diff
@@ -15,6 +18,16 @@ mod imp {
+        #[template_child]
+        pub(crate) verticalspace_respect_vertical_bordersrow: TemplateChild<adw::SwitchRow>,
+        #[template_child]
+        pub(crate) verticalspace_respect_horizontal_bordersrow: TemplateChild<adw::SwitchRow>,
```
I'd rename these fields in the following way: `verticaltool` -> `verticalspace` and respect_<horizontal/vertical>_bordersrow` -> `limit_movement_<horizontal/vertical>_borders_row`

flxzt commented on crates/rnote-compose/src/lib.rs:
```diff
@@ -31,6 +31,8 @@ pub mod style;
 /// other misc utilities
 pub mod utils;
+/// vertical tool options
+pub mod verticaltoolconfig;
```
Why is this in rnote-compose? I think it should be in rnote-engine under `pensconfig` (and renamed to verticalspacetoolconfig`

Doublonmousse committed:
Delete ._toolspage.ui

Doublonmousse committed:
renamings

Doublonmousse committed:
put back the vertical tool config struct inside `rnote-engine`

Doublonmousse committed:
clean up `keys_between`

Doublonmousse committed:
fix option types and renames

Doublonmousse commented:
The CI check is giving warnings for a line that doesn't exist ?
```
 warning: unused import: `tracing::error`
  --> crates/rnote-engine/src/store/stroke_comp.rs:15:5
   |
15 | use tracing::error;
   |     ^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default
```
But there's nothing 
https://github.com/Doublonmousse/rnote/blob/vertical_tool/crates/rnote-engine/src/store/stroke_comp.rs#L15

flxzt commented:
I noticed it as well. Seems to be a false positive lint.

One additional point, which I am not sure about:

Should strokes that only intersect the limits be included as well or not?

Doublonmousse commented:
I think it's better to have all selection tools work when the full stroke is in bounds, and add a global option for partial selection like #155 for all tools.

Do you think adding a graphical indicator when limiting the selection horizontally to show where the selection stops is a good idea ? I also wondered if I should do something in that case to make it harder to accidentally overlap strokes onto the next page (like some snapping or indicator when the selection would touch the next horizontal border)

flxzt commented:
At least from my testing it was intuitive enough without an indicator.

Snapping sound like a good idea, I can see it being useful often

Doublonmousse committed:
use the `rtree` to find strokes in `keys_between`

Doublonmousse commented:
I'm not sure snapping is a super good idea. I don't know how to deal with potentially two snapping modes at the same time (General snapping to grid + the added one when crossing the border)

Maybe it makes more sense to limit the amplitude of the tool so that the selected portion stays in the same page instead. This way I don't have the issue with multiple snapping modes, and we don't end up with strokes that become unmovable by the vertical space tool when using the limiter on horizontal borders (because of the decision to only include strokes that are fully in bounds).

flxzt commented:
Couldn't there be both? Just like content already also snaps to the page borders when `Snap positions` is enabled.

flxzt commented:
But implementing snapping is also not a blocker for me

Doublonmousse commented:
Oh I see, that's indeed possible to have two snapping modes.

That would make placing strokes between pages when using the horizontal limit less of an issue.

Maybe that can be a follow up ? I wonder if it's better to select strokes that intersect the bounds in this PR in the meantime

flxzt commented:
> Maybe that can be a follow up ? I wonder if it's better to select strokes that intersect the bounds in this PR in the meantime

Of course.

I just tested it again, and including intersecting strokes feels more intuitive, so I'd say yes.

But I found one remaining issue: When both limits are enabled in the UI, the vertical option actually is not respected.
There seems to be a logic error in the implementation

Doublonmousse committed:
fix logic + selection of strokes that intersect the bounds

Doublonmousse committed:
fix logic

Doublonmousse commented:
I think it's working now

flxzt commented:
Looks good now, nice!
